### PR TITLE
[iOS] Wrap HTTPS-by-default logic around feature flags

### PR DESCRIPTION
**Pull request by kylehickinson**
Resolves https://github.com/brave/brave-browser/issues/40634

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by Brandon-T — APPROVED** [review]

File: ios/browser/flags/about_flags.mm
Comment (ShivanKaul) [inline ios/browser/flags/about_flags.mm:90]: I see an "HTTPS Upgrades" feature flag which is default disabled on iOS, how does that relate to this one?

File: ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+WKNavigationDelegate.swift
Comment (ShivanKaul) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+WKNavigationDelegate.swift:1834]: Where do we fallback to our old way of upgrading to HTTPS i.e. via WebKit's internal mechanism?

Comment (kylehickinson) [reply ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+WKNavigationDelegate.swift:1834]: We do so when creating the web view here: https://github.com/brave/brave-core/blob/31c975b2ccc270c22e332f28d862029be5a44aa6/ios/brave-ios/Sources/Brave/Frontend/Browser/Tab.swift#L493

The above still matches the prior behaviour which was based on the old preference that looked like this:
```swift
configuration!.upgradeKnownHostsToHTTPS = Preferences.Shields.httpsEverywhere.value
```

File: ios/browser/flags/about_flags.mm
Comment (kylehickinson) [reply ios/browser/flags/about_flags.mm:90]: These control the similar Chromium iOS features, which are only supported in Chromium web views https://source.chromium.org/chromium/chromium/src/+/main:ios/components/security_interstitials/https_only_mode/feature.h

The goal will be to use these in the future either in this PR https://github.com/brave/brave-core/pull/24657 or after it merges

**Review by StephenHeaps — APPROVED** [review]

**Review by iefremov — APPROVED** [review]
chromium_src ++

**Review by Brandon-T — APPROVED** [review]

**Comment by ShivanKaul** [comment]
Just an update here that I'm chatting with @diracdeltas and @fmarier about whether we want this, let's not merge for now.

**Review by arthuredelstein — APPROVED** [review]

**Review by ShivanKaul — APPROVED** [review]

**Comment by kjozwiak** [comment]
Verification **PASSED** on `iPhone 11` running `iOS 17.5.1` via the following build(s):

```
Brave | 1.71.29 Chromium: 128.0.6613.85 (Official Build) nightly (64-bit) 
--- | ----
Revision | c532a6657223...
OS | iOS
```

### `Test Case #1` - `brave://flags#https-only-mode` enabled (default mode)

* ensured that `brave://flags#https-only-mode` is set as enabled by default on a clean/new install
* ensured `Upgrade Connections to HTTPS` is set as `Standard` as the default

**`Example`** | **`Example`**
----------------|------------------
IMG_0468 | IMG_0471

While both `brave://flags#https-by-default` & `brave://flags#https-only-mode` are enabled, went through the STR/Cases outlined via https://github.com/brave/brave-browser/issues/36408#issue-2157693231 as per the following:

**`Test interstitial`**

**`Example`** | **`Example`** | **`Example`** | **`Example`** | **`Example`** 
----------------|----------------|----------------|----------------|---------------
IMG_0472 | IMG_0473 | IMG_0474 | IMG_0475 | IMG_0476

**`Test upgrading`**

**`Example`** | **`Example`** | **`Example`** | **`Example`** | **`Example`** | **`Example`** 
----------------|----------------|----------------|----------------|----------------|-----------------
IMG_0479 | IMG_0480 | IMG_0481 | IMG_0482 | IMG_0483 | IMG_0484

### `Test Case #2` - `brave://flags#https-only-mode` disabled

* disabled `brave://flags#https-only-mode` via `brave://flags` after installing `1.71.29 Chromium: 128.0.6613.85`
* restart the browser once the `brave://flag` has been changed
* once disabled, ensured that `Upgrade Connections to HTTPS` is set as a boolean toggle (enabled/disabled)
* visited http://http.badssl.com and ensured that a interstitial page re: HTTP upgrade is **NOT** being displayed

**`Example`** | **`Example`** | **`Example`** 
----------------|----------------|----------------
IMG_0486 | IMG_0487 | IMG_0488

### `Test Case #3` - `brave://flags#https-by-default` disabled

* disabled `brave://flags#https-by-default` via `brave://flags` after installing `1.71.29 Chromium: 128.0.6613.85`
* restart the browser once the `brave://flag` has been changed
* ensured that `Upgrade Connections to HTTPS` is set as a boolean toggle (enabled/disabled)
   * `brave://flags#https-only-mode` is also being disabled when `brave://flags#https-by-default` has been disabled 
* visited http://http.badssl.com and ensured that a interstitial page re: HTTP upgrade is **NOT** being displayed
* ensured that `http://brave.com` --> `https://brave.com`
* ensured that `http://google.com` --> `https://google.com`
* ensured that `http://facebook.com` --> `https://m.facebook.com`
* ensured that `http://reddit.com` --> `https://reddit.com`

**`Example`** | **`Example`** | **`Example`** 
----------------|----------------|----------------
IMG_0489 | IMG_0490 | IMG_0491

**Pull request merged**
